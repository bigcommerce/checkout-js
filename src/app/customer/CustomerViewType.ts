enum CustomerViewType {
    Guest = 'guest',
    Login = 'login',
    SuggestedLogin = 'suggested_login',
    EnforcedLogin = 'enforced_login',
    CancellableEnforcedLogin = 'cancellable_enforced_login',
}

export default CustomerViewType;
