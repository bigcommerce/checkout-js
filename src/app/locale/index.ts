import { WithCurrencyProps } from './withCurrency';
import { WithLanguageProps } from './withLanguage';
import { LocaleContextType } from './LocaleContext';
import { TranslatedHtmlProps } from './TranslatedHtml';
import { TranslatedLinkProps } from './TranslatedLink';
import { TranslatedStringProps } from './TranslatedString';

export type LocaleContextType = LocaleContextType;
export type WithCurrencyProps = WithCurrencyProps;
export type WithLanguageProps = WithLanguageProps;
export type TranslatedHtmlProps = TranslatedHtmlProps;
export type TranslatedStringProps = TranslatedStringProps;
export type TranslatedLinkProps = TranslatedLinkProps;

export { default as LocaleContext } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getLanguageService } from './getLanguageService';
export { default as withCurrency } from './withCurrency';
export { default as withLanguage } from './withLanguage';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml } from './TranslatedHtml';
export { default as TranslatedLink } from './TranslatedLink';
export { default as TranslatedString } from './TranslatedString';
