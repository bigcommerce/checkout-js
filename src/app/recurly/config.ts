export const recurlyId = window.RECURLY_PUBLIC_KEY ?? process.env.RECURLY_PUBLIC_KEY;

// @ts-ignore
export const apiEndpoint = window.FRONTEND_API_ENDPOINT ?? process.env.FRONTEND_API_ENDPOINT;
// @ts-ignore
export const recurlyBcAppId = window.RECURLY_BC_APP_ID ?? process.env.RECURLY_BC_APP_ID;
export default {recurlyId, apiEndpoint, recurlyBcAppId};
