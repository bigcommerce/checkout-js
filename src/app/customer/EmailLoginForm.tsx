import { SignInEmail } from '@bigcommerce/checkout-sdk';
import { withFormik, FormikProps } from 'formik';
import { noop } from 'lodash';
import React, { memo, useCallback, FunctionComponent } from 'react';

import { withLanguage, TranslatedString, WithLanguageProps } from '../locale';
import { Alert, AlertType } from '../ui/alert';
import { Button, ButtonVariant } from '../ui/button';
import { Form } from '../ui/form';
import { LoadingSpinner } from '../ui/loading';
import { Modal, ModalHeader } from '../ui/modal';

import getEmailValidationSchema from './getEmailValidationSchema';
import EmailField from './EmailField';

export interface EmailLoginFormProps {
    email?: string;
    isOpen: boolean;
    isSendingEmail?: boolean;
    emailHasBeenRequested?: boolean;
    sentEmail?: SignInEmail;
    sentEmailError?: any;
    onRequestClose?(): void;
    onSendLoginEmail?(values: EmailLoginFormValues): void;
}

export interface EmailLoginFormValues {
    email: string;
}

const EmailLoginForm: FunctionComponent<EmailLoginFormProps & WithLanguageProps & FormikProps<EmailLoginFormValues>> = ({
    email,
    isOpen,
    isSendingEmail = false,
    emailHasBeenRequested,
    onRequestClose,
    sentEmailError,
    sentEmail,
    values: {
        email: formEmail,
    },
}) => {
    const modalHeaderStringId = () => {
        if (emailHasBeenRequested) {
            if (sentEmailError) {
                return 'common.error_heading';
            }

            return 'login_email.sent_header';
        }

        if (email) {
            return 'login_email.header_with_email';
        }

        return 'login_email.header';

    };

    const renderFooter = useCallback(() => (
        <div className="modal-footer">
            { sentEmailError && sentEmailError.status === 429 ?
                <Button
                    onClick={ onRequestClose }
                    type="button"
                    variant={ ButtonVariant.Primary }
                >
                    <TranslatedString id="common.ok_action" />
                </Button> :
                <>
                    <Button
                        className="optimizedCheckout-buttonSecondary"
                        onClick={ onRequestClose }
                        type="button"
                    >
                        <TranslatedString id="common.cancel_action" />
                    </Button>
                    <Button
                        isLoading={ isSendingEmail }
                        type="submit"
                        variant={ ButtonVariant.Primary }
                    >
                        <TranslatedString id={ emailHasBeenRequested ? 'login_email.resend' : 'login_email.send' } />
                    </Button>
                </> }
        </div>
    ), [sentEmailError, emailHasBeenRequested, isSendingEmail, onRequestClose]);

    const renderError = useCallback(() => {
        if (!sentEmailError) {
            return null;
        }

        const { status } = sentEmailError;

        return (
            <Alert type={ AlertType.Error }>
                { status === 429 ?
                    <TranslatedString id="login_email.error_temporary_disabled" /> :
                    <TranslatedString id={ status === 404 ?
                        'login_email.error_not_found' :
                        'login_email.error_server' }
                    /> }
            </Alert>
        );
    }, [sentEmailError]);

    const renderForm = useCallback(() => {
        if (sentEmailError && sentEmailError.status === 429) {
            return null;
        }

        if (emailHasBeenRequested && sentEmail && !sentEmailError) {
            const { expiry, sent_email } = sentEmail;

            return (
                <p>
                    <TranslatedString
                        data={ {
                            email: formEmail,
                            minutes: Math.round(expiry / 60),
                        } }
                        id={ sent_email === 'sign_in' ?
                            'login_email.sent_text' :
                            'customer.reset_password_before_login_error' }
                    />
                </p>
            );
        }

        if (emailHasBeenRequested && !sentEmail) {
            return <EmailField />;
        }

        return (<>
            <p>
                <TranslatedString id="login_email.text" />
            </p>
            <EmailField />
        </>);
    }, [sentEmailError, emailHasBeenRequested, sentEmail, formEmail]);

    return (
        <Modal
            additionalBodyClassName="modal--withText"
            additionalModalClassName="modal--medium"
            header={
                <ModalHeader>
                    <TranslatedString id={ modalHeaderStringId() } />
                </ModalHeader>
            }
            isOpen={ isOpen }
            onRequestClose={ onRequestClose }
            shouldShowCloseButton={ true }
        >
            <Form>
                <LoadingSpinner isLoading={ isSendingEmail && !email } />
                { renderError() }
                { renderForm() }
                { renderFooter() }
            </Form>
        </Modal>);
};

export default withLanguage(withFormik<EmailLoginFormProps & WithLanguageProps, EmailLoginFormValues>({
    mapPropsToValues: ({
        email = '',
    }) => ({
        email,
    }),
    handleSubmit: (values, { props: { onSendLoginEmail = noop } }) => {
        onSendLoginEmail(values);
    },
    validationSchema: ({ language }: WithLanguageProps) => getEmailValidationSchema({ language }),
})(memo(EmailLoginForm)));
