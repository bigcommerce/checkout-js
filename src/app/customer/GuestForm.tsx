import { withFormik, FormikProps } from 'formik';
import React, { memo, FunctionComponent, ReactNode } from 'react';
import { object, string } from 'yup';

import { withLanguage, TranslatedHtml, TranslatedString, WithLanguageProps } from '../locale';
import { Button, ButtonVariant } from '../ui/button';
import { BasicFormField, Fieldset, Form, Legend  } from '../ui/form';

import EmailField from './EmailField';
import SubscribeField from './SubscribeField';

export interface GuestFormProps {
    canSubscribe: boolean;
    checkoutButtons?: ReactNode;
    defaultShouldSubscribe: boolean;
    email?: string;
    isContinuingAsGuest: boolean;
    onChangeEmail(email: string): void;
    onContinueAsGuest(data: GuestFormValues): void;
    onShowLogin(): void;
}

export interface GuestFormValues {
    email: string;
    shouldSubscribe: boolean;
}

const GuestForm: FunctionComponent<GuestFormProps & WithLanguageProps & FormikProps<GuestFormValues>> = ({
    canSubscribe,
    checkoutButtons,
    isContinuingAsGuest,
    onChangeEmail,
    onShowLogin,
}) => (
    <Form
        className="checkout-form"
        id="checkout-customer-guest"
        testId="checkout-customer-guest"
    >
        <Fieldset
            legend={
                <Legend hidden>
                    <TranslatedString id="customer.guest_customer_text" />
                </Legend>
            }
        >
            <p>
                <TranslatedHtml id="customer.checkout_as_guest_text" />
            </p>

            <div className="customerEmail-container">
                <div className="customerEmail-body">
                    <EmailField onChange={ onChangeEmail } />

                    { canSubscribe && <BasicFormField
                        component={ SubscribeField }
                        name="shouldSubscribe"
                    /> }
                </div>

                <div className="form-actions customerEmail-action">
                    <Button
                        className="customerEmail-button"
                        id="checkout-customer-continue"
                        isLoading={ isContinuingAsGuest }
                        testId="customer-continue-as-guest-button"
                        type="submit"
                        variant={ ButtonVariant.Primary }
                    >
                        <TranslatedString id="customer.continue_as_guest_action" />
                    </Button>
                </div>
            </div>

            <p>
                <TranslatedString id="customer.login_text" />
                { ' ' }
                <a
                    data-test="customer-continue-button"
                    id="checkout-customer-login"
                    onClick={ onShowLogin }
                >
                    <TranslatedString id="customer.login_action" />
                </a>
            </p>

            { checkoutButtons }
        </Fieldset>
    </Form>
);

export default withLanguage(withFormik<GuestFormProps & WithLanguageProps, GuestFormValues>({
    mapPropsToValues: ({
        email = '',
        defaultShouldSubscribe = false,
    }) => ({
        email,
        shouldSubscribe: defaultShouldSubscribe,
    }),
    handleSubmit: (values, { props: { onContinueAsGuest } }) => {
        onContinueAsGuest(values);
    },
    validationSchema: ({ language }: GuestFormProps & WithLanguageProps) => {
        const email = string()
            .email(language.translate('customer.email_invalid_error'))
            .max(256)
            .required(language.translate('customer.email_required_error'));

        return object({ email });
    },
})(memo(GuestForm)));
