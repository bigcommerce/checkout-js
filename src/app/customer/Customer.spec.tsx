import { createCheckoutService, BillingAddress, Checkout, CheckoutService, Customer as CustomerData, StoreConfig } from '@bigcommerce/checkout-sdk';
import { mount, render, ReactWrapper } from 'enzyme';
import React, { FunctionComponent } from 'react';

import { getBillingAddress } from '../billing/billingAddresses.mock';
import { CheckoutProvider } from '../checkout';
import { getCheckout } from '../checkout/checkouts.mock';
import { getStoreConfig } from '../config/config.mock';
import { createLocaleContext, LocaleContext, LocaleContextType } from '../locale';

import { getCustomer, getGuestCustomer } from './customers.mock';
import Customer, { CustomerProps, WithCheckoutCustomerProps } from './Customer';
import CustomerViewType from './CustomerViewType';
import EmailLoginForm from './EmailLoginForm';
import GuestForm, { GuestFormProps } from './GuestForm';
import LoginForm, { LoginFormProps } from './LoginForm';

describe('Customer', () => {
    let CustomerTest: FunctionComponent<CustomerProps & Partial<WithCheckoutCustomerProps>>;
    let billingAddress: BillingAddress;
    let checkout: Checkout;
    let checkoutService: CheckoutService;
    let config: StoreConfig;
    let customer: CustomerData;
    let localeContext: LocaleContextType;

    beforeEach(() => {
        billingAddress = getBillingAddress();
        checkout = getCheckout();
        customer = getGuestCustomer();
        config = getStoreConfig();

        checkoutService = createCheckoutService();

        jest.spyOn(checkoutService.getState().data, 'getBillingAddress')
            .mockReturnValue(billingAddress);

        jest.spyOn(checkoutService.getState().data, 'getCheckout')
            .mockReturnValue(checkout);

        jest.spyOn(checkoutService.getState().data, 'getCustomer')
            .mockReturnValue(customer);

        jest.spyOn(checkoutService.getState().data, 'getConfig')
            .mockReturnValue(config);

        localeContext = createLocaleContext(getStoreConfig());

        CustomerTest = props => (
            <CheckoutProvider checkoutService={ checkoutService }>
                <LocaleContext.Provider value={ localeContext }>
                    <Customer { ...props } />
                </LocaleContext.Provider>
            </CheckoutProvider>
        );
    });

    describe('when view type is "guest"', () => {
        it('matches snapshot', () => {
            expect(render(
                <CustomerTest viewType={ CustomerViewType.Guest } />
            ))
                .toMatchSnapshot();
        });

        it('renders guest form by default', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Guest } />
            );

            expect(component.find(GuestForm).exists())
                .toEqual(true);
        });

        it('renders guest form if billing address is undefined', () => {
            jest.spyOn(checkoutService.getState().data, 'getBillingAddress')
                .mockReturnValue(undefined);

            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Guest } />
            );

            expect(component.find(GuestForm).exists())
                .toEqual(true);
        });

        it('renders guest form if customer is undefined', () => {
            jest.spyOn(checkoutService.getState().data, 'getCustomer')
                .mockReturnValue(undefined);

            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Guest } />
            );

            expect(component.find(GuestForm).exists())
                .toEqual(true);
        });

        it('passes data to guest form', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Guest } />
            );

            expect(component.find(GuestForm).props())
                .toMatchObject({
                    canSubscribe: config.shopperConfig.showNewsletterSignup,
                    defaultShouldSubscribe: config.shopperConfig.defaultNewsletterSignup,
                    email: billingAddress.email,
                });
        });

        it('continues checkout as guest and does not send consent if not required', () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const component = mount(
                <CustomerTest
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: ' test@bigcommerce.com ',
                    shouldSubscribe: true,
                });

            expect(checkoutService.continueAsGuest)
                .toHaveBeenCalledWith({
                    email: 'test@bigcommerce.com',
                    acceptsMarketingNewsletter: true,
                    acceptsAbandonedCartEmails: true,
                });
        });

        it('only subscribes to newsletter if allowed by customer', () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const subscribeToNewsletter = jest.fn();
            const component = mount(
                <CustomerTest
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: false,
                });

            expect(checkoutService.continueAsGuest)
                .toHaveBeenCalledWith({
                    email: 'test@bigcommerce.com',
                    acceptsMarketingNewsletter: undefined,
                    acceptsAbandonedCartEmails: undefined,
                });

            expect(subscribeToNewsletter)
                .not.toHaveBeenCalled();
        });

        it('changes to login view when "show login" event is received', () => {
            const handleChangeViewType = jest.fn();
            const component = mount(
                <CustomerTest
                    onChangeViewType={ handleChangeViewType }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onShowLogin')();

            expect(handleChangeViewType)
                .toHaveBeenCalledWith(CustomerViewType.Login);
        });

        it('triggers completion callback if customer successfully continued as guest', async () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const handleContinueAsGuest = jest.fn();
            const component = mount(
                <CustomerTest
                    onContinueAsGuest={ handleContinueAsGuest }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: false,
                });

            await new Promise(resolve => process.nextTick(resolve));

            expect(handleContinueAsGuest)
                .toHaveBeenCalled();
        });

        it('triggers completion callback if continueAsGuest fails with update_subscriptions', async () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockRejectedValue({ type: 'update_subscriptions' });

            const handleContinueAsGuest = jest.fn();
            const component = mount(
                <CustomerTest
                    onContinueAsGuest={ handleContinueAsGuest }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: true,
                });

            await new Promise(resolve => process.nextTick(resolve));

            expect(handleContinueAsGuest)
                .toHaveBeenCalled();
        });

        it('renders CancellableEnforcedLogin if continue as guest fails with code 403', async () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockRejectedValue({ status: 403 });

            const handleChangeViewType = jest.fn();
            const component = mount(
                <CustomerTest
                    onChangeViewType={ handleChangeViewType }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: false,
                });

            await new Promise(resolve => process.nextTick(resolve));
            component.update();

            expect(handleChangeViewType).toHaveBeenCalledWith(CustomerViewType.CancellableEnforcedLogin);
        });

        it('renders SuggestedLogin form if continue as guest returns truthy shouldEncourageSignIn', async () => {
            jest.spyOn(checkoutService.getState().data, 'getCustomer')
                .mockReturnValue({
                    ...getCustomer(),
                    isGuest: true,
                    shouldEncourageSignIn: true,
                } as any);

            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const handleChangeViewType = jest.fn();
            const component = mount(
                <CustomerTest
                    onChangeViewType={ handleChangeViewType }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: false,
                });

            await new Promise(resolve => process.nextTick(resolve));
            component.update();

            expect(handleChangeViewType).toHaveBeenCalledWith(CustomerViewType.SuggestedLogin);

            (checkoutService.getState().data.getCustomer as jest.Mock).mockRestore();
        });

        it('renders EnforcedLogin form if continue as guest fails with code 429', async () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockRejectedValue({ status: 429 });

            const handleChangeViewType = jest.fn();

            const component = mount(
                <CustomerTest
                    onChangeViewType={ handleChangeViewType }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: false,
                });

            await new Promise(resolve => process.nextTick(resolve));
            component.update();

            expect(handleChangeViewType).toHaveBeenCalledWith(CustomerViewType.EnforcedLogin);
        });

        it('triggers error callback if customer is unable to continue as guest', async () => {
            jest.spyOn(checkoutService, 'continueAsGuest')
                .mockRejectedValue({ type: 'unknown_error' });

            const handleError = jest.fn();
            const component = mount(
                <CustomerTest
                    onContinueAsGuestError={ handleError }
                    viewType={ CustomerViewType.Guest }
                />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onContinueAsGuest')({
                    email: 'test@bigcommerce.com',
                    shouldSubscribe: false,
                });

            await new Promise(resolve => process.nextTick(resolve));

            expect(handleError)
                .toHaveBeenCalled();
        });

        it('retains draft email address when switching to login view', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Guest } />
            );

            (component.find(GuestForm) as ReactWrapper<GuestFormProps>)
                .prop('onChangeEmail')('test@bigcommerce.com');

            component.setProps({ viewType: CustomerViewType.Login });
            component.update();

            expect((component.find(LoginForm) as ReactWrapper<LoginFormProps>).prop('email'))
                .toEqual('test@bigcommerce.com');
        });
    });

    describe('when view type is "login"', () => {
        it('matches snapshot', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Login } />
            );

            expect(component.render())
                .toMatchSnapshot();
        });

        it('renders login form', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Login } />
            );

            expect(component.find(LoginForm).exists())
                .toEqual(true);
        });

        it('renders sign-in email when link is clicked', async () => {
            const component = mount(
                <CustomerTest
                    isSignInEmailEnabled={ true }
                    viewType={ CustomerViewType.Login }
                />);

            expect(component.find(EmailLoginForm).exists())
                .toEqual(false);

            component.find('[data-test="customer-signin-link"]').simulate('click');
            await new Promise(resolve => process.nextTick(resolve));

            expect(component.find(EmailLoginForm).prop('emailHasBeenRequested'))
                .toEqual(false);
        });

        it('passes data to login form', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Login } />
            );

            expect(component.find(LoginForm).props())
                .toMatchObject({
                    email: billingAddress.email,
                    canCancel: config.checkoutSettings.guestCheckoutEnabled,
                    createAccountUrl: config.links.createAccountLink,
                    forgotPasswordUrl: config.links.forgotPasswordLink,
                });
        });

        it('handles "sign in" event', () => {
            jest.spyOn(checkoutService, 'signInCustomer')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Login } />
            );

            (component.find(LoginForm) as ReactWrapper<LoginFormProps>)
                .prop('onSignIn')({
                    email: 'test@bigcommerce.com',
                    password: 'password1',
                });

            expect(checkoutService.signInCustomer)
                .toHaveBeenCalledWith({
                    email: 'test@bigcommerce.com',
                    password: 'password1',
                });
        });

        it('triggers completion callback if customer is successfully signed in', async () => {
            jest.spyOn(checkoutService, 'signInCustomer')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const handleSignedIn = jest.fn();
            const component = mount(
                <CustomerTest
                    onSignIn={ handleSignedIn }
                    viewType={ CustomerViewType.Login }
                />
            );

            (component.find(LoginForm) as ReactWrapper<LoginFormProps>)
                .prop('onSignIn')({
                    email: 'test@bigcommerce.com',
                    password: 'password1',
                });

            await new Promise(resolve => process.nextTick(resolve));

            expect(handleSignedIn)
                .toHaveBeenCalled();
        });

        it('triggers error callback if customer is unable to sign in', async () => {
            jest.spyOn(checkoutService, 'signInCustomer')
                .mockRejectedValue({ type: 'unknown_error' });

            const handleError = jest.fn();
            const component = mount(
                <CustomerTest
                    onSignInError={ handleError }
                    viewType={ CustomerViewType.Login }
                />
            );

            (component.find(LoginForm) as ReactWrapper<LoginFormProps>)
                .prop('onSignIn')({
                    email: 'test@bigcommerce.com',
                    password: 'password1',
                });

            await new Promise(resolve => process.nextTick(resolve));

            expect(handleError)
                .toHaveBeenCalled();
        });

        it('clears error when "cancel" event is triggered', () => {
            const error = new Error();

            jest.spyOn(checkoutService.getState().errors, 'getSignInError')
                .mockReturnValue(error);

            jest.spyOn(checkoutService, 'clearError')
                .mockReturnValue(Promise.resolve(checkoutService.getState()));

            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Login } />
            );

            (component.find(LoginForm) as ReactWrapper<LoginFormProps>)
                // tslint:disable-next-line:no-non-null-assertion
                .prop('onCancel')!();

            expect(checkoutService.clearError)
                .toHaveBeenCalledWith(error);
        });

        it('changes to guest view when "cancel" event is triggered', () => {
            const handleChangeViewType = jest.fn();
            const component = mount(
                <CustomerTest
                    onChangeViewType={ handleChangeViewType }
                    viewType={ CustomerViewType.Login }
                />
            );

            (component.find(LoginForm) as ReactWrapper<LoginFormProps>)
                // tslint:disable-next-line:no-non-null-assertion
                .prop('onCancel')!();

            expect(handleChangeViewType)
                .toHaveBeenCalledWith(CustomerViewType.Guest);
        });

        it('retains draft email address when switching to guest view', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.Login } />
            );

            (component.find(LoginForm) as ReactWrapper<LoginFormProps>)
                // tslint:disable-next-line:no-non-null-assertion
                .prop('onChangeEmail')!('test@bigcommerce.com');

            component.setProps({ viewType: CustomerViewType.Guest });
            component.update();

            expect((component.find(GuestForm) as ReactWrapper<GuestFormProps>).prop('email'))
                .toEqual('test@bigcommerce.com');
        });
    });

    describe('when view type is "cancellable_enforced_login"', () => {
        it('renders login form', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.CancellableEnforcedLogin } />
            );

            expect(component.find(LoginForm).prop('viewType'))
                .toEqual(CustomerViewType.CancellableEnforcedLogin);
        });
    });

    describe('when view type is "suggested_login"', () => {
        it('renders login form', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.SuggestedLogin } />
            );

            expect(component.find(LoginForm).prop('viewType'))
                .toEqual(CustomerViewType.SuggestedLogin);
        });
    });

    describe('when view type is "enforced_login"', () => {
        it('renders login form', () => {
            const component = mount(
                <CustomerTest viewType={ CustomerViewType.EnforcedLogin } />
            );

            expect(component.find(LoginForm).prop('viewType'))
                .toEqual(CustomerViewType.EnforcedLogin);
        });

        it('calls sendLoginEmail and renders form when sign-in email link is clicked ', async () => {
            const sendLoginEmail = jest.fn(() => new Promise(resolve => resolve()) as any);
            const component = mount(
                <CustomerTest
                    email="foo@bar.com"
                    isSignInEmailEnabled={ true }
                    sendLoginEmail={ sendLoginEmail }
                    viewType={ CustomerViewType.EnforcedLogin }
                />);

            component.find('[data-test="customer-signin-link"]').simulate('click');
            component.update();
            await new Promise(resolve => process.nextTick(resolve));
            component.update();

            expect(sendLoginEmail).toHaveBeenCalledWith('foo@bar.com');
            expect(component.find(EmailLoginForm).prop('emailHasBeenRequested'))
                .toEqual(true);
        });

        it('renders EmailLoginForm even when sendLoginForm is rejected', async () => {
            const sendLoginEmail = jest.fn(() => new Promise((_, reject) => reject()) as any);
            const component = mount(
                <CustomerTest
                    email="foo@bar.com"
                    isSignInEmailEnabled={ true }
                    sendLoginEmail={ sendLoginEmail }
                    viewType={ CustomerViewType.EnforcedLogin }
                />);

            component.find('[data-test="customer-signin-link"]').simulate('click');
            component.update();
            await new Promise(resolve => process.nextTick(resolve));
            component.update();

            expect(sendLoginEmail).toHaveBeenCalledWith('foo@bar.com');
            expect(component.find(EmailLoginForm).prop('emailHasBeenRequested'))
                .toEqual(true);
        });
    });
});
