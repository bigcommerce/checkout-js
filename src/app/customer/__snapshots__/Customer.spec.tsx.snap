// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer when view type is "guest" matches snapshot 1`] = `
<div>
  <div>
    <form
      class="checkout-form"
      data-test="checkout-customer-guest"
      id="checkout-customer-guest"
      novalidate=""
    >
      <fieldset
        class="form-fieldset"
      >
        <legend
          class="form-legend is-srOnly"
        >
          Guest Customer
        </legend>
        <div
          class="form-body"
        >
          <p>
            <span>
              Checking out as a 
              <strong>
                Guest
              </strong>
              ? You'll be able to save your details to create an account with us later.
            </span>
          </p>
          <div
            class="customerEmail-container"
          >
            <div
              class="customerEmail-body"
            >
              <div
                class="form-field"
              >
                <label
                  class="form-label optimizedCheckout-form-label"
                  for="email"
                >
                  Email Address
                </label>
                <input
                  autocomplete="email"
                  class="form-input optimizedCheckout-form-input"
                  id="email"
                  name="email"
                  type="email"
                  value="test@bigcommerce.com"
                />
              </div>
              <div
                class="form-field"
              >
                <input
                  class="form-checkbox"
                  id="shouldSubscribe"
                  name="shouldSubscribe"
                  type="checkbox"
                  value="false"
                />
                <label
                  class="form-label optimizedCheckout-form-label"
                  for="shouldSubscribe"
                >
                  Subscribe to our newsletter.
                </label>
              </div>
            </div>
            <div
              class="form-actions customerEmail-action"
            >
              <button
                class="button customerEmail-button button--primary optimizedCheckout-buttonPrimary"
                data-test="customer-continue-as-guest-button"
                id="checkout-customer-continue"
                type="submit"
              >
                Continue as guest
              </button>
            </div>
          </div>
          <p>
            Already have an account? 
            <a
              data-test="customer-continue-button"
              id="checkout-customer-login"
            >
              Sign in now
            </a>
          </p>
        </div>
      </fieldset>
    </form>
  </div>
</div>
`;

exports[`Customer when view type is "login" matches snapshot 1`] = `
<div>
  <div>
    <form
      class="checkout-form"
      data-test="checkout-customer-returning"
      id="checkout-customer-returning"
      novalidate=""
    >
      <fieldset
        class="form-fieldset"
      >
        <legend
          class="form-legend is-srOnly"
        >
          Returning Customer
        </legend>
        <div
          class="form-body"
        >
          <p>
            Don't have an account? 
            <a
              href="#"
            >
              Create an account
            </a>
             to continue.
          </p>
          <div
            class="form-field"
          >
            <label
              class="form-label optimizedCheckout-form-label"
              for="email"
            >
              Email Address
            </label>
            <input
              autocomplete="email"
              class="form-input optimizedCheckout-form-input"
              id="email"
              name="email"
              type="email"
              value="test@bigcommerce.com"
            />
          </div>
          <div
            class="form-field"
          >
            <label
              class="form-label optimizedCheckout-form-label"
              for="password"
            >
              Password
            </label>
            <div
              class="form-field-password"
            >
              <input
                class="form-input optimizedCheckout-form-input form-input--withIcon"
                id="password"
                name="password"
                type="password"
                value=""
              />
              <a
                class="form-toggle-password form-input-icon"
                href="#"
              >
                <div
                  class="icon"
                >
                  <svg
                    viewBox="0 0 640 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M634 471L36 3.51A16 16 0 0 0 13.51 6l-10 12.49A16 16 0 0 0 6 41l598 467.49a16 16 0 0 0 22.49-2.49l10-12.49A16 16 0 0 0 634 471zM296.79 146.47l134.79 105.38C429.36 191.91 380.48 144 320 144a112.26 112.26 0 0 0-23.21 2.47zm46.42 219.07L208.42 260.16C210.65 320.09 259.53 368 320 368a113 113 0 0 0 23.21-2.46zM320 112c98.65 0 189.09 55 237.93 144a285.53 285.53 0 0 1-44 60.2l37.74 29.5a333.7 333.7 0 0 0 52.9-75.11 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64c-36.7 0-71.71 7-104.63 18.81l46.41 36.29c18.94-4.3 38.34-7.1 58.22-7.1zm0 288c-98.65 0-189.08-55-237.93-144a285.47 285.47 0 0 1 44.05-60.19l-37.74-29.5a333.6 333.6 0 0 0-52.89 75.1 32.35 32.35 0 0 0 0 29.19C89.72 376.41 197.08 448 320 448c36.7 0 71.71-7.05 104.63-18.81l-46.41-36.28C359.28 397.2 339.89 400 320 400z"
                    />
                  </svg>
                </div>
              </a>
            </div>
            <a
              data-test="forgot-password-link"
              href="https://store-k1drp8k8.bcapp.dev/login.php?action=reset_password"
              rel="noopener noreferrer"
              target="_blank"
            >
              Forgot password?
            </a>
          </div>
          <div
            class="form-actions"
          >
            <button
              class="button button--primary optimizedCheckout-buttonPrimary"
              data-test="customer-continue-button"
              id="checkout-customer-continue"
              type="submit"
            >
              Sign In
            </button>
            <a
              class="button optimizedCheckout-buttonSecondary"
              data-test="customer-cancel-button"
              href="#"
              id="checkout-customer-cancel"
            >
              Cancel
            </a>
          </div>
        </div>
      </fieldset>
    </form>
  </div>
</div>
`;
